pkg/ip/ipmasq_nftables_linux.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ip
+
+import (
+	"context"
+	"net"
+
+	"sigs.k8s.io/knftables"
+
+	"github.com/containernetworking/plugins/pkg/utils"
+)
+
+const (
+	ipMasqTableName = "cni_plugins_masquerade"
+	ipMasqChainName = "masq_checks"
+)
+
+// The nftables ipmasq implementation is basically a straight port of the iptables
+// implementation: we add a rule for each mapping, with a comment containing a hash of the
+// container ID, so that we can later reliably delete the rules we want. (This is
+// important because in edge cases, it's possible the plugin might see "ADD container A
+// with IP 192.168.1.3", followed by "ADD container B with IP 192.168.1.3" followed by
+// "DEL container A with IP 192.168.1.3", and we need to make sure that the DEL causes us
+// to delete the rule for container A, and not the rule for container B.)
+//
+// It would be more nftables-y to have a chain with a single rule doing a lookup against a
+// set with an element per mapping, rather than having a chain with a rule per mapping.
+// But there's no easy, non-racy way to say "delete the element 192.168.1.3 from the set,
+// but only if it was added for container A, not if it was added for container B".
+
+// hashForContainer returns a unique hash identifying the rules for this container with
+// this plugin
+func hashForContainer(pluginName, containerID string) string {
+	return utils.MustFormatHashWithPrefix(16, "", pluginName+":"+containerID)
+}
+
+// setupIPMasqNFTables is the nftables-based implementation of SetupIPMasq
+func setupIPMasqNFTables(ipn *net.IPNet, pluginName, containerID string) error {
+	nft, err := knftables.New(knftables.InetFamily, ipMasqTableName)
+	if err != nil {
+		return err
+	}
+	return setupIPMasqNFTablesWithInterface(nft, ipn, pluginName, containerID)
+}
+
+func setupIPMasqNFTablesWithInterface(nft knftables.Interface, ipn *net.IPNet, pluginName, containerID string) error {
+	comment := hashForContainer(pluginName, containerID)
```

Comment:
The iptables implementation is wrong here -- we need to be able to find all masquerade rules for a given network. Can the comment include network name in plain text?